### PR TITLE
Thumbnails: save dashboard thumbnails in SQL

### PR DESCRIPTION
**Pull request by ryantxu**
:warning: running this will add a table to SQL.. be sure it is right before merging

Followup to #43226 that saves images in SQL rather than on disk

### CODE REVIEW

File: pkg/services/thumbs/crawler.go [hunk 1]
```diff
@@ -29,19 +34,21 @@ type simpleCrawler struct {
...
+	store  *sqlstore.SQLStore
 	mode   CrawlerMode
 	opts   rendering.Opts
 	status crawlStatus
 	queue  []dashItem
 	mu     sync.Mutex
 }
 
-func newSimpleCrawler(folder string, renderService rendering.Service, gl *live.GrafanaLive) dashRenderer {
+func newSimpleCrawler(folder string, renderService rendering.Service, gl *live.GrafanaLive, store *sqlstore.SQLStore) dashRenderer {
```
Comment (ArturWierzbicki) [inline pkg/services/thumbs/crawler.go:45]: should I depend on `sqlstore.SQLStore` and use it directly or should I send commands to it through the `bus`? if the latter, does the same rule apply to `rendering.Service` and `live.GrafanaLive`? @ryantxu @FZambia

File: pkg/models/dashboard_thumbs.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	return ThumbnailKindDefault, errors.New("unknown thumbnail kind " + str)
+}
+
+// A DashboardThumbnail includes all metadata for a dashboard thumbnail
+type DashboardThumbnail struct {
+	Id           int64         `json:"id"`
+	DashboardId  int64         `json:"dashboardId"`
+	PanelId      int64         `json:"panelId,omitempty"`
+	Kind         ThumbnailKind `json:"kind"`
+	Theme        string        `json:"theme"` // TODO: changing it from `string` to `rendering.Theme` causes gen-go to fail with undescriptive error
```
Comment (ArturWierzbicki) [inline pkg/models/dashboard_thumbs.go:44]: I am not sure whats going on here, `wire` handles `ThumbnailKind` just fine but fails with `Theme`

Comment (ryantxu) [reply pkg/models/dashboard_thumbs.go:44]: :shrug: maybe @FZambia has some guesses?

File: pkg/services/thumbs/crawler.go (hunk 1 above)
Comment (ryantxu) [reply pkg/services/thumbs/crawler.go:45]: we should avoid using `bus` where possible -- now that we can inject explicit interfaces (wire) we are slowly removing bus usage so the real dependencies are more discoverable.

File: pkg/services/thumbs/crawler.go [hunk 3]
```diff
@@ -247,3 +237,56 @@ func (r *simpleCrawler) walk() {
 	r.status.Finished = time.Now()
 	r.broadcastStatus()
 }
+
+// TODO extract the three methods below from this file
+func (r *simpleCrawler) SaveThumbnailFromFile(tempFilePath string, dashboardID int64, dashboardUID string, theme rendering.Theme, kind models.ThumbnailKind) (int64, error) {
```
Comment (ryantxu) [inline pkg/services/thumbs/crawler.go:242]:
```suggestion
func (r *simpleCrawler) saveThumbnailFromFile(tempFilePath string, dashboardID int64, dashboardUID string, theme rendering.Theme, kind models.ThumbnailKind) (int64, error) {
```

Does this need to be public?  better to keep things private (lower case) unless really necessary

File: pkg/services/thumbs/crawler.go [hunk 4]
```diff
@@ -247,3 +237,56 @@ func (r *simpleCrawler) walk() {
...
+
+	if err != nil {
+		tlog.Error("error reading file", "dashboardUID", dashboardUID, "err", err)
+		return 0, err
+	}
+
+	return r.SaveThumbnailFromBytes(content, r.GetMimeType(tempFilePath), dashboardID, dashboardUID, theme, kind)
+}
+
+func (r *simpleCrawler) GetMimeType(filePath string) string {
```
Comment (ryantxu) [inline pkg/services/thumbs/crawler.go:267]:
```suggestion
func getMimeType(filePath string) string {
```
no need for this to be a member function (and you can remove it from the interface)

File: pkg/services/thumbs/models.go [hunk 5]
```diff
@@ -93,15 +50,17 @@ type crawlStatus struct {
...
 
 	// Assumes you have already authenticated as admin
 	Stop() (crawlStatus, error)
 
 	// Assumes you have already authenticated as admin
 	Status() (crawlStatus, error)
+
+	SaveThumbnailFromFile(tempFilePath string, dashboardID int64, dashboardUID string, theme rendering.Theme, kind models.ThumbnailKind) (int64, error)
+	SaveThumbnailFromBytes(content []byte, mimeType string, dashboardID int64, dashboardUID string, theme rendering.Theme, kind models.ThumbnailKind) (int64, error)
+	GetMimeType(filePath string) string
```
Comment (ryantxu) [inline pkg/services/thumbs/models.go:65]: these feel like internal details that should not be exported or exposed on the interface.

The 'save' functions feel like they belong elsewhere, but assuming they are internal I think they are OK for now

File: public/app/features/search/components/SearchCard.tsx [hunk 6]
```diff
@@ -18,15 +18,40 @@ export interface Props {
   onToggleChecked?: OnToggleChecked;
 }
 
-export function getThumbnailURL(uid: string, isLight?: boolean) {
-  return `/api/dashboards/uid/${uid}/img/thumb/${isLight ? 'light' : 'dark'}`;
-}
+export const getThumbnailURL = (uid: string, isLight?: boolean) =>
```
Comment (ryantxu) [inline public/app/features/search/components/SearchCard.tsx:21]: I know this is just a style thing... but what is wrong with exporting simple functions!

Comment (ArturWierzbicki) [reply public/app/features/search/components/SearchCard.tsx]: haha I just touched the method and my IDE autoconverted it - I'll disable that

File: pkg/services/thumbs/models.go (hunk 5 above)
Comment (ArturWierzbicki) [reply pkg/services/thumbs/models.go]: :+1:

File: pkg/models/dashboard_thumbs.go (hunk 2 above)
Comment (FZambia) [reply pkg/models/dashboard_thumbs.go]: rendering package import models package already - so it's due to cycle import most probably

**Comment by ArturWierzbicki** [comment]
closing in favor of https://github.com/grafana/grafana/pull/44083 with all sql changes

**Pull request closed without merging**